pallets/fee-lock/src/lib.rs

```diff
@@ -60,7 +60,50 @@ pub mod pallet {
 	pub struct Pallet<T>(PhantomData<T>);
 
 	#[pallet::hooks]
-	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {}
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+		fn on_idle(now: T::BlockNumber, remaining_weight: Weight) -> Weight {
+			let mut consumed_weight: Weight = Default::default();
+
+			// process only up to 80% or remaining weight
+			let cost_of_single_unlock = T::WeightInfo::unlock_fee() // cost of unlock action
+				+ T::DbWeight::get().reads(1)   // Self::get_fee_lock_metadata
+				+ T::DbWeight::get().reads(1); // single iteration
+
+			let cost_of_single_unlock_iteration = T::WeightInfo::unlock_fee() // cost of unlock action
+				+ T::DbWeight::get().reads(1); // iteration
+
+			if cost_of_single_unlock.ref_time() > remaining_weight.ref_time() {
+				return Weight::from_ref_time(0)
+			}
+
+			consumed_weight += T::DbWeight::get().reads(1);
+			let metadata = Self::get_fee_lock_metadata();
+			let period_length = metadata.map(|meta| meta.period_length);
+			let mut accounts_it = AccountFeeLockData::<T>::iter();
+
+			loop {
+				consumed_weight += T::DbWeight::get().reads(1);
+				match (period_length, accounts_it.next()) {
+					(Some(period_lenght), Some((who, lock))) => {
+						let unlock_block = lock.last_fee_lock_block.checked_add(&period_lenght);
+
+						if matches!(unlock_block, Some(unlock) if unlock <= now) {
+							consumed_weight += T::WeightInfo::unlock_fee();
+							let _ = <Self as FeeLockTriggerTrait<T::AccountId>>::unlock_fee(&who);
+						}
+					},
+					_ => break,
+				};
+
+				if cost_of_single_unlock_iteration.ref_time() >
+					(remaining_weight.ref_time() - consumed_weight.ref_time())
```

Comment:
Agree, we should probably at least add a warning log on all `break` cases.

Reply:
> Feels a bit risky. I understand it will work correctly with the current impl but any modifications later on will make it panic. 

What exactly is risky here, actually i dont think its possible to make to working wrong, if accounts for every manual storage read/write & uses the bench marked weights for extrinsics.

So to make it not work you would have to modify on_idle hook itself, then its your responsibility to take care of things,  i also came up with a lot of unit tests for all the important corner cases i guess

```
running 20 tests
test mock::__construct_runtime_integrity_test::runtime_integrity_tests ... ok
test tests::test_on_idle_unlock_multiple_users::unlocks_tokens_for_single_account_only_with_limited_weight ... ok
test tests::test_autounlock_on_empty_unlock_queue ... ok
test tests::test_on_idle_unlock_for_single_user::does_not_unlock_tokens_when_weigh_is_zero ... ok
test tests::test_on_idle_unlock_for_single_user::unlock_tokens_using_a_bit_more_weight_that_required ... ok
test tests::test_on_idle_unlock_for_single_user::unlocks_tokens_for_an_user ... ok
test tests::test_unlock_happens_not_sooner_but_after_period ... ok
test tests::test_on_idle_unlock_multiple_users::unlocks_tokens_for_both_users_using_exact_required_weight_ ... ok
test tests::test_on_idle_unlock_multiple_users::unlocks_tokens_for_both_users_with_unlimited_input_weight ... ok
test tests::test_maintain_queue_with_subsequent_fee_locks_on_single_account ... ok
test tests::test_process_queue_and_ignore_outdated_items_in_unlock_queue_because_of_subsequent_process_fee_lock_calls ... ok
test tests::process_fee_lock_trigger_works ... ok
test tests::test_process_queue_and_ignore_outdated_items_in_unlock_queue_because_of_manual_unlock ... ok
test tests::test_unlock_happens_in_order ... ok
test tests::test_on_idle_unlock_for_single_user::unlock_tokens_using_a_too_small_weight_that_required ... ok
test tests::test_on_idle_unlock_for_single_user::unlock_tokens_using_exact_amount_of_weight_required ... ok
test tests::test_unlock_stops_after_single_iteration_without_consuming_unnecessary_weight ... ok
test tests::update_fee_lock_metadata_works ... ok
test tests::whitelist_and_valuation_works ... ok
test tests::unlock_fee_works ... ok

```

Reply:
If the tests cover it and we are not modifying the hook without due considerations, then its fine